src/cpu_treeshap_model.h

```diff
@@ -0,0 +1,358 @@
+{
+    float res = 0.;
+    for (int i = 0; i < d; i++)
+    {
+        res += e[i] *offset[i] / (Base[i] + q) * n[i];
+    }
+    return res / d;
+}
+
+void times(const float *input, float *output, float scalar, int size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        output[i] = input[i] * scalar;
+    }
+};
+
+void times_broadcast(const float *input, float *output, int size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        output[i] *= input[i];
+    }
+};
+
+void addition(float *input, float *output, int size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        output[i] += input[i];
+    }
+};
+
+void write(float *from, float *to, int size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        to[i] = from[i];
+    }
+};
+
+
+
+template <typename ThresholdType>
+bool decision_non_categorical(float fvalue, ThresholdType threshold, treelite::Operator op) {
+  // XGBoost
+  if (op == treelite::Operator::kLT) {
+    return fvalue < threshold;
+  }
+  // LightGBM, sklearn, cuML RF
+  if (op == treelite::Operator::kLE) {
+    return fvalue <= threshold;
+  }
+  switch (op) {
+    case treelite::Operator::kEQ:
+      return fvalue == threshold;
+    case treelite::Operator::kGT:
+      return fvalue > threshold;
+    case treelite::Operator::kGE:
+      return fvalue >= threshold;
+    default:
+      TREELITE_CHECK(false) << "Unrecognized comparison operator " << static_cast<int>(op);
+      return -1;
+  }
+}
+
+inline bool decision_categorical(float fvalue, const std::vector<std::uint32_t>& matching_categories,
+                               bool categories_list_right_child) {
+  bool is_matching_category;
+  auto max_representable_int = static_cast<float>(std::uint32_t(1) << FLT_MANT_DIG);
+  if (fvalue < 0 || std::fabs(fvalue) > max_representable_int) {
+    is_matching_category = false;
+  } else {
+    const auto category_value = static_cast<std::uint32_t>(fvalue);
+    is_matching_category = (
+        std::find(matching_categories.begin(), matching_categories.end(), category_value)
+        != matching_categories.end());
+  }
+  if (categories_list_right_child) {
+    return !is_matching_category;
+  } else {
+    return is_matching_category;
+  }
+}
+
+template <typename tl_threshold_t, typename tl_output_t>
+bool
+decision(
+    const treelite::Tree<tl_threshold_t, tl_output_t>& tree, const float fvalue,
+    int node)
+{
+  // Missing
+  if(std::isnan(fvalue)){
+    return tree.DefaultChild(node) == tree.LeftChild(node);
+  }
+
+  if (tree.SplitType(node) == treelite::SplitFeatureType::kCategorical) {
+    return decision_categorical(
+        fvalue, tree.MatchingCategories(node),
+        tree.CategoriesListRightChild(node));
+  } else {
+    return decision_non_categorical(fvalue, tree.Threshold(node), tree.ComparisonOp(node));
+  }
+}
+
+
+template <typename tl_threshold_t, typename tl_output_t>
+void inference(const TreeMetaInfo<tl_threshold_t, tl_output_t> &tree_info,
```

Comment:
The linear treeshap algorithm implementation looks good! 

One catch is that currently, here, for simplicity of implementation, the max_depth is picked at max 32, and trees with different depths will reuse the top tree_depth sub-elements of `kBase` from 32 Chebyshev points of the second kind.  This is mathematically correct but might not enjoy the numerical stability as depth varied  Chebyshev points selection.

Reply:
If I use different points for each depth I think I have to repeat the kNorm matrix 32 times right? It uses too much space.

Reply:
Yes, I'm not saying you should do that. But just something to keep in mind :) when you set the overflow tolerance in the test.